Proposition de mise à jour du wsdl

dlevasseuraltitude opened the pull request:
Voir Readme.md avec patchnote

dlevasseuraltitude committed:
Mise à jour wsdl initial

Corrections diverses

dlevasseuraltitude committed:
Update du Readme suite modification wsdl

Ajout du patchnote v0.2

dlevasseuraltitude committed:
Merge pull request #1 from dlevasseuraltitude/patch-1

Update du Readme suite modification wsdl

aminvielledebat commented on README.md:
```diff
@@ -1 +1,9 @@
+  - Suppession des ArrayOf... car tout les tableaux sont gérés via des séquences
+     - Les "ArrayOf" faisait que nous ne pouvions pas retourner un tableau avec un seul élément
+  - Suppression des "maxLength" sur les <xs:element>
+     - Le wsdl n'était pas valide avec ces attributs
```
Hello !
Cool pour la proposition. 
Pour ce qui est du README, on fonctionnerait comment du coup ? Chaque PR porte sa modif, ou on le fait au moment du tag ?

aminvielledebat commented:
Closes #2 #1

dlevasseuraltitude committed:
Modification du wsdl

Modification pour correction de trois éléments problématiques: 
- Il manquait l'attribut "unbounded" sur le ListeFibreType
- Les entiers qui avaient pour type "unsignedInt" ont été remplacés par des "nonNegativeInteger".
- Le type "EtatFibreType" est maintenant un énumérateur pour respecter la documentation

dlevasseuraltitude committed:
Update README.md

Ajout des informations de modification du wsdl

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
\ No newline at end of file
+	<wsdl:types>
+		<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" targetNamespace="http://interop-fibre.fr/">
+			<xs:import namespace="http://schemas.datacontract.org/2004/07/"/>
```
Je ne veux pas dire de bétise, mais de mémoire ce namespace n'existe pas, c'est juste une convention pour WCF (.NET).

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+			<xs:element name="RecherchePBO">
+				<xs:complexType>
+					<xs:sequence>
+						<xs:element xmlns:q1="http://schemas.datacontract.org/2004/07/" minOccurs="0" name="demande" nillable="true" type="q1:RecherchePBODemande"/>
```
Il faudrait spécifier le Data Contract Namespace en précisant un domaine intérop plutôt.
D'ailleurs pourquoi utiliser les WCF pour un wsdl qui doit être multi language ?

De ce que j'ai vu/lu (sauf erreur, car je ne maitrise pas trop WCF) ce n'est jamais évident d'utiliser ça pour certains langage et c'est très orienté .NET... soit pas du tout interopérable.

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
 <wsdl:definitions xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:wsap="http://schemas.xmlsoap.org/ws/2004/08/addressing/policy" xmlns:wsa10="http://www.w3.org/2005/08/addressing" xmlns:tns="http://interop-fibre.fr/" xmlns:msc="http://schemas.microsoft.com/ws/2005/12/wsdl/contract" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:wsx="http://schemas.xmlsoap.org/ws/2004/09/mex"  ...
```
Il manque toujours le versionning de namespace qui est super important.
J'ai fait une PR dans ce sens. Tu peux t'appuyer dessus ou la valider et merge/rebase ton travail (~5mins max).

dlevasseuraltitude replied:
Ok j'ajoute cette modification dans le fichier WSDL

aminvielledebat replied:
Super ! Merci beaucoup :).
Si tu as besoin d'aide, j'en remets une couche, mais nous pouvons vous aider sur le sujet.

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+			<xs:complexType name="RecherchePBODemande">
+				<xs:sequence>
+					<xs:element name="Entete" nillable="true" type="tns:EnteteRequeteType"/>
+					<xs:element minOccurs="0" name="ReferenceCommandePriseInterneOC" nillable="true" type="xs:string"/>
```
Pourquoi ne pas avoir ajouté une longueur max comme tu l'avais souhaité au départ ?
Je rappelle l'ancien code : 

```xml
<xs:element minOccurs="0" maxLength="20" name="ReferenceCommandePriseInterneOC" nillable="true"
            type="xs:string"/>
<xs:element minOccurs="0" maxLength="20" name="ReferencePrestationPrise" nillable="true"
            type="xs:string"/>
```
La perte de la longueur max risque de nous amener les mêmes problèmes qu'avec les fichiers interop', à savoir une gestion spécifique par ORTs.
De plus c'est décrit explicitement dans la spécification Word. Il faut donc l'implémenter.

De notre côté, on peut vous aider sur le sujet, mais il faut se mettre d'accord.

dlevasseuraltitude replied:
Alors de notre côté, en le passant dans un validateur, ces éléments étaient retournées comme étant en erreur. Je vais me pencher là dessus voir si j'ai un moyen de les intégrer sans erreur.

aminvielledebat replied:
Je te confirme, l'attribute ```maxLength``` n'est pas conforme. Il faut utiliser des restrictions en définissant les différents types de la façon suivante : 

```xml
<xsd:simpleType name="TonTypeQuiVaBien">
        <xsd:restriction base="xsd:string">
            <xsd:minLength value="LeMinDeCars"/>
            <xsd:maxLength value="LeMaxDeCars"/>
        </xsd:restriction>
    </xsd:simpleType>
```

Bon par contre, c'est clair que c'est long et fastidieux :-(

dlevasseuraltitude replied:
Ok merci pour l'info, je regarde de mon côté pour intégrer ces restrictions

dlevasseuraltitude replied:
Par exemple ça donnerai ça pour l'élément "ReferenceCommandePriseInterneOC" tu confirmes?
```
<xs:element minOccurs="0" name="ReferenceCommandePriseInterneOC" nillable="true" type="xs:string">
        <xs:simpleType>
	        <xs:restriction base="xs:string">
		        <xs:minLength value="0"/>
		        <xs:maxLength value="20"/>
	        </xs:restriction>
        </xs:simpleType>
</xs:element>
```

aminvielledebat replied:
Yes ! Modulo le ```minOccurs``` (je pense que le champ est obligatoire ne l’occurrence) et le ```nillable``` (est ce que l'on permet l'envoi d'un champ avec la balise ```nil=true``` et que doit-on considérer si on définit le ```nillable``` dans le xsd mais que l'OC/OI ne l'envoie pas (car minOccurs = 0)).

Un avis @jlsaxione  ?

jlsaxione replied:
juste pour info, le client ne se génère pas du tout en java. j'attends la prochaine version.
Est il possible d'avoir accès au repo personnel  ?juste pour tenter de générer avec les évolutions ?

jlsaxione replied:
Cette référence d'un point de vue fonctionnel est elle obligatoire ? dans la pdc elle l'est.
dans ce cas le  minLength à 0 et le nillable true me gène.

aminvielledebat replied:
Pour l'accès à la version en cours, je suis du même avis que @jlsaxione. 
Le plus simple serait que vous pushiez sur une branche à vous sur ce repo. Cela nous permettrait au passage de pouvoir se baser pour nos PRs (ex avec la mienne sur la pagination, où je vais devoir faire un rebase suite à votre merge).

jlsaxione replied:
bon avec un plugin maven cxf le code se génère, en revanche le code généré est loin d'être propre.
Je rejoins Alexis sur le sujet des encapsulations  org.datacontract et com.microsoft.

aminvielledebat replied:
Géré dans la #21

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+				<xs:sequence>
+					<xs:element name="Entete" nillable="true" type="tns:EnteteRequeteType"/>
+					<xs:element minOccurs="0" name="ReferenceCommandePriseInterneOC" nillable="true" type="xs:string"/>
+					<xs:element minOccurs="0" name="ReferencePrestationPrise" nillable="true" type="xs:string"/>
```
Idem et sur tous les autres pour lesquels vous avez supprimé le contrôle de longueur.

dlevasseuraltitude replied:
Idem que mon commentaire précédent

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
-<wsdl:part name="parameters" element="tns:MiseAJourRouteOptiqueResponse"/>
-</wsdl:message>
-<wsdl:portType name="Emutation">
-<wsdl:documentation>WebServices de mutation made by Corentin Le Guen (corentin.leguen@altitudeinfra.fr)</wsdl:documentation>
```
C'est dommage d'avoir supprimé la documentation (même si justifié ici :-)) sur le service et les opérations.
Il faudrait la rajouter. Pourquoi ne pas copier/coller la description du word Interop' ?

•	RecherchePBO : fourniture du PBO affecté à la commande ;PBO de rattachement pour un accès déjà en service et dans le cadre d’un SAV (PBO courant) et les éventuels PBO voisins.
•	ConfirmeAdresse : Liste les adresses de la ZA du PBO choisi et confirmation ou modification de l'adresse fournie à la commande et en SAV
•	ConsultationFibres : consultation de l’ensemble des fibres d'un PBO avec leur état,
•	MiseAJourRouteOptique : modification de la route optique et fourniture d'un N° de décharge

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+			<xs:element name="MiseAJourRouteOptiqueReponse" nillable="true" type="tns:MiseAJourRouteOptiqueReponse"/>
+			<xs:complexType name="ListeRoutesOptiques">
+				<xs:sequence>
+					<xs:element name="RoutesOptiques" nillable="true" type="tns:RouteOptique"/>
```
Il manque le maxOccurs, sinon ta collection ne fera qu'un seul élément.

dlevasseuraltitude replied:
Bien vu, mais lors d'une mise à jour de route optique, n'est-on pas censé retourner une seule route optique. Dans quels cas peut-on en remonter plusieurs? Pour nous il s'agit d'une erreur dans la documentation

aminvielledebat replied:
De ce que j'ai cru comprendre dans la doc, on parle de 1 à 4 fibres en fonction du type d'ingénierie, comme dans le Cr ou la Notif de Reprov.
En phase avec toi sur le peu d'utilité de l'information. C'est quelque chose que nous avons déjà fait remonter plusieurs fois en interop'.

dlevasseuraltitude replied:
Effectivement en relisant la doc le type ListeRouteOptique retourne jusqu'à quatre routes, je vais donc faire la modification, par contre lorsque l'on met à jour une route, on réserve une seule fibre, qui pour moi ne correspond forcément qu'à une seule route. Il faudrait en rediscuter en groupe interop car pour moi retourner plus d'une route n'a pas sens.

aminvielledebat replied:
Complètement en phase ! Mais c'est comme le Cr et la NotifReprov...

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+					<xsd:enumeration value="fibre dédiée disponible"/>
+					<xsd:enumeration value="fibre réservée"/>
+					<xsd:enumeration value="fibre occupée"/>
+					<xsd:enumeration value="fibre hors service"/>
```
:+1:

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+			<xs:element name="ConsultationFibres">
+				<xs:complexType>
+					<xs:sequence>
+						<xs:element xmlns:q3="http://schemas.datacontract.org/2004/07/" minOccurs="0" name="demande" nillable="true" type="q3:ConsultationFibresDemande"/>
```
Je ne comprends pas à quoi sert ce niveau d'encapsulation supplémentaire. Peux-tu nous expliquer ?

Selon moi, il faudrait mieux le supprimer si il ne sert pas.
i.e : 
- Pour les objets requêtes, tu crées systématiquement un élément ```demande```, qui plus est nillable qui n'a pas lieu d'être (à quoi sert-il ? ou alors je n'ai pas compris).
- Pour les objets réponses, même chose avec un sous élément ```NomOperation**Result**```, et même remarque sur le fait qu'il soit nillable.

dlevasseuraltitude replied:
Je suis d'accord concernant le nillable, qui n'a pas d'intérêt ici. Pour moi cette encapsulation sert à faire le lien entre la méthode et le type complexe passé en paramètre dans le cas de la demande, et le type complexe de retour dans le cas de la réponse.

aminvielledebat replied:
Tu pourras m'expliquer pour l'encapsulation ? Au pire on peut s'appeler.
Comme ça, je dirai que le seul avantage est de permettre le pattern middleware, mais à part ça :confused: ... normalement tu peux mapper directement, sans avoir à ajouter ce niveau supplémentaire.

Avec les exemples qui vont bien. Avec tes définitions, ma requête de base donne ça :
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:int="http://interop-fibre.fr/" xmlns:ns="http://schemas.datacontract.org/2004/07/">
   <soapenv:Header>
      </soapenv:Header>
   <soapenv:Body>
      <int:RecherchePBO>
         <!--Optional:-->
         <int:demande>
            <ns:Entete>
               <ns:HorodatageRequete>?</ns:HorodatageRequete>
               <ns:OperateurCommercial>
                  <!--Optional:-->
                  <ns:Identifiant>?</ns:Identifiant>
                  <ns:Nom>?</ns:Nom>
               </ns:OperateurCommercial>
               <ns:VersionWS>?</ns:VersionWS>
            </ns:Entete>
            <!--Optional:-->
            <ns:ReferenceCommandePriseInterneOC>?</ns:ReferenceCommandePriseInterneOC>
            <!--Optional:-->
            <ns:ReferencePrestationPrise>?</ns:ReferencePrestationPrise>
         </int:demande>
      </int:RecherchePBO>
   </soapenv:Body>
</soapenv:Envelope>
```
Alors que dans mon esprit, on était plus censé avoir : 
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:int="http://interop-fibre.fr/" xmlns:ns="http://schemas.datacontract.org/2004/07/">
   <soapenv:Header>
      </soapenv:Header>
   <soapenv:Body>
      <int:RecherchePBO>
            <ns:Entete>
               <ns:HorodatageRequete>?</ns:HorodatageRequete>
               <ns:OperateurCommercial>
                  <!--Optional:-->
                  <ns:Identifiant>?</ns:Identifiant>
                  <ns:Nom>?</ns:Nom>
               </ns:OperateurCommercial>
               <ns:VersionWS>?</ns:VersionWS>
            </ns:Entete>
            <!--Optional:-->
            <ns:ReferenceCommandePriseInterneOC>?</ns:ReferenceCommandePriseInterneOC>
            <!--Optional:-->
            <ns:ReferencePrestationPrise>?</ns:ReferencePrestationPrise>
      </int:RecherchePBO>
   </soapenv:Body>
</soapenv:Envelope>
```

dlevasseuraltitude replied:
Effectivement, sans cette encapsulation ça marche aussi, donc je vais l'enlever du wsdl

aminvielledebat replied:
Top ! Merci :+1:

dlevasseuraltitude replied:
Alors après des tests de mon coté, le WS est fonctionnel sans cette encapsulation, mais quand j'essaie de l'intégrer dans mon appli C#, j'ai un problème avec les types complexes. De mon côté impossible donc de l'intégrer sans cette encapsulation. Je vais donc MAJ le wsdl avec toutes les autres modifications, après peut-être que tu verras quelque chose qui m'aurait échappé pour enlever l'encapsulation

aminvielledebat replied:
Une histoire de fous ça. Tu peux nous mettre un copier/coller du wsdl/xsd sans l'encapsulation ici stp ?
On peut peut être identifier d'où vient le problème avec un oeil neuf.

dlevasseuraltitude replied:
Voilà ce que j'ai de mon côté:

```
<wsdl:definitions xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:wsap="http://schemas.xmlsoap.org/ws/2004/08/addressing/policy" xmlns:wsa10="http://www.w3.org/2005/08/addressing" xmlns:tns="http://interop-fibre.fr/Emutation/1.0/" xmlns:msc="http://schemas.microsoft.com/ws/2005/12/wsdl/contract" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:wsx="http://schemas.xmlsoap.org/ws/2004/09/mex" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:wsam="http://www.w3.org/2007/05/addressing/metadata" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsp="http://schemas.xmlsoap.org/ws/2004/09/policy" xmlns:wsaw="http://www.w3.org/2006/05/addressing/wsdl" xmlns:soap12="http://schemas.xmlsoap.org/wsdl/soap12/" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:xsd="http://www.w3.org/2001/XMLSchema" name="Emutation" targetNamespace="http://interop-fibre.fr/Emutation/1.0/">
	<wsdl:types>
		<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" targetNamespace="http://interop-fibre.fr/Emutation/1.0/">
			<xs:element name="RecherchePBO" element="tns:RecherchePBO">
				<xs:complexType>
					<xs:sequence>
						<xs:element name="Entete" nillable="true" type="tns:EnteteRequeteType"/>
						<xs:element minOccurs="0" name="ReferenceCommandePriseInterneOC" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="ReferencePrestationPrise" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
					</xs:sequence>
				</xs:complexType>
			</xs:element>
			<xs:element name="RecherchePBOResponse" element="tns:RecherchePBOResponse">
				<xs:complexType>
					<xs:sequence>
						<xs:element name="CodeRetour" nillable="true" type="tns:CodeRetourType"/>
						<xs:element name="Entete" nillable="true" type="tns:EnteteReponseType"/>
						<xs:element name="ListePBO" nillable="true" type="tns:ListePboType"/>
					</xs:sequence>
				</xs:complexType>
			</xs:element>
			<xs:element name="ConsultationFibres" element="tns:ConsultationFibres">
				<xs:complexType>
					<xs:sequence>
						<xs:element name="Entete" nillable="true" type="tns:EnteteRequeteType"/>
						<xs:element minOccurs="0" name="ReferenceCommandePriseInterneOC" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="ReferencePBO" nillable="true" type="xs:string"/>
						<xs:element minOccurs="0" name="ReferencePM" nillable="true" type="xs:string"/>
						<xs:element minOccurs="0" name="ReferencePrestationPrise" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
					</xs:sequence>
				</xs:complexType>
			</xs:element>
			<xs:element name="ConsultationFibresResponse" element="tns:ConsultationFibresResponse">
				<xs:complexType>
					<xs:sequence>
						<xs:element name="CodeRetour" nillable="true" type="tns:CodeRetourType"/>
						<xs:element name="Entete" nillable="true" type="tns:EnteteReponseType"/>
						<xs:element name="Fibre" nillable="true" type="tns:ListeFibreType"/>
						<xs:element name="PboType" nillable="true" type="tns:PboType"/>
					</xs:sequence>
				</xs:complexType>
			</xs:element>
			<xs:element name="MiseAJourRouteOptique" element="tns:MiseAJourRouteOptique">
				<xs:complexType>
					<xs:sequence>
						<xs:element name="Batiment" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="60"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="BatimentTerrain" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="60"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="CodeRetour" nillable="true" type="tns:CodeRetourType"/>
						<xs:element name="Entete" nillable="true" type="tns:EnteteRequeteType"/>
						<xs:element name="Escalier" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="25"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="EscalierTerrain" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="25"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="Etage" nillable="true" >
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="25"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="EtageTerrain" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="25"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="IdentifiantFibre" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="30"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="MotifMutation" nillable="true" type="xs:string"/>
						<xs:element minOccurs="0" name="PorteTerrain" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="25"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="ReferenceAdresse" nillable="true" type="tns:ReferenceAdresseDemandeType"/>
						<xs:element minOccurs="0" name="ReferenceCommandePriseInterneOC" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="ReferencePrestationPrise" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element minOccurs="0" name="ReferencePrise" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
					</xs:sequence>
				</xs:complexType>
			</xs:element>
			<xs:element name="MiseAJourRouteOptiqueResponse"  element="tns:MiseAJourRouteOptiqueResponse">
				<xs:complexType>
					<xs:sequence>
						<xs:element name="CodeRetour" nillable="true" type="tns:CodeRetourType"/>
						<xs:element name="Entete" nillable="true" type="tns:EnteteReponseType"/>
						<xs:element name="NumeroDecharge" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="ReferencePBO" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="30"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="ReferencePM" nillable="true" type="xs:string"/>
						<xs:element name="ReferencePMT" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="20"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="ReferencePrise" nillable="true">
							<xs:simpleType>
								<xs:restriction base="xs:string">
								  <xs:minLength value="0"/>
								  <xs:maxLength value="30"/>
								</xs:restriction>
							</xs:simpleType>
						</xs:element>
						<xs:element name="RoutesOptiques" nillable="true" type="tns:ListeRoutesOptiques"/>
					</xs:sequence>
				</xs:complexType>
			</xs:element>
		</xs:schema>
		<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:tns="http://interop-fibre.fr/Emutation/1.0/" elementFormDefault="qualified" targetNamespace="http://interop-fibre.fr/Emutation/1.0/">
			<xs:complexType name="EnteteRequeteType">
				<xs:sequence>
					<xs:element name="HorodatageRequete" type="xs:dateTime"/>
					<xs:element name="OperateurCommercial" nillable="true" type="tns:OperateurCommercialType"/>
					<xs:element name="VersionWS" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="EnteteRequeteType" nillable="true" type="tns:EnteteRequeteType"/>
			<xs:complexType name="OperateurCommercialType">
				<xs:sequence>
					<xs:element minOccurs="0" name="Identifiant" nillable="true" type="xs:string"/>
					<xs:element name="Nom" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="OperateurCommercialType" nillable="true" type="tns:OperateurCommercialType"/>
			<xs:complexType name="CodeRetourType">
				<xs:sequence>
					<xs:element minOccurs="0" name="CodeErreur" nillable="true" type="xs:string"/>
					<xs:element name="CodeRetour" type="xs:int"/>
					<xs:element minOccurs="0" name="LibelleErreur" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="CodeRetourType" nillable="true" type="tns:CodeRetourType"/>
			<xs:complexType name="EnteteReponseType">
				<xs:sequence>
					<xs:element name="HorodatageReponse" type="xs:dateTime"/>
					<xs:element name="HorodatageRequete" type="xs:dateTime"/>
					<xs:element name="IdentifiantReponse" type="xs:int"/>
					<xs:element name="OperateurCommercial" nillable="true" type="tns:OperateurCommercialType"/>
					<xs:element name="VersionWS" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="EnteteReponseType" nillable="true" type="tns:EnteteReponseType"/>
			<xs:complexType name="ListePboType">
				<xs:sequence>
					<xs:element minOccurs="0" maxOccurs="unbounded" name="Pbo" nillable="true" type="tns:PboType"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="ListePboType" nillable="true" type="tns:ListePboType"/>
			
			<xs:complexType name="PboType">
				<xs:sequence>
					<xs:element name="AdressePBO" nillable="true" type="tns:AdressePBOReponseType"/>
					<xs:element minOccurs="0" name="NaturePBO" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="NombreFibresDisponibles" type="xs:int"/>
					<xs:element name="ReferencePBO" nillable="true" type="xs:string"/>
					<xs:element name="ReferencePM"  nillable="true" type="xs:string"/>
					<xs:element name="StructurePBO" nillable="true" type="tns:StructureVerticalePBOType"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="PboType" nillable="true" type="tns:PboType"/>
			<xs:complexType name="AdressePBOReponseType">
				<xs:sequence>
					<xs:element minOccurs="0" name="AdressePBO" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferenceGeographique" nillable="true" type="tns:CoordonneesGeographiquesType"/>
					<xs:element minOccurs="0" name="ReferenceHexacle" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferenceHexacleVoie" nillable="true" type="tns:ReferenceHexacleVoieType"/>
					<xs:element minOccurs="0" name="ReferenceRivoli" nillable="true" type="tns:ReferenceRivoliType"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="AdressePBOReponseType" nillable="true" type="tns:AdressePBOReponseType"/>
			<xs:complexType name="CoordonneesGeographiquesType">
				<xs:sequence>
					<xs:element minOccurs="0" name="CoordonneesX" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="CoordonneesY" nillable="true" type="xs:string"/>
					<xs:element name="TypeProjection" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="CoordonneesGeographiquesType" nillable="true" type="tns:CoordonneesGeographiquesType"/>
			<xs:complexType name="ReferenceHexacleVoieType">
				<xs:sequence>
					<xs:element name="CodeHexacleVoie"  nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ComplementNumeroVoie" type="xs:string"/>
					<xs:element name="NumeroVoie" type="xs:nonNegativeInteger"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="ReferenceHexacleVoieType" nillable="true" type="tns:ReferenceHexacleVoieType"/>
			<xs:complexType name="ReferenceRivoliType">
				<xs:sequence>
					<xs:element name="CodeInsee"  nillable="true" type="xs:string"/>
					<xs:element name="CodeRivoli"  nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ComplementNumeroVoie" type="xs:string" />
					<xs:element name="NumeroVoie" type="xs:nonNegativeInteger"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="ReferenceRivoliType" nillable="true" type="tns:ReferenceRivoliType"/>
			<xs:complexType name="StructureVerticalePBOType">
				<xs:sequence>
					<xs:element name="Batiment" nillable="true" type="xs:string"/>
					<xs:element name="Escalier" nillable="true" type="xs:string"/>
					<xs:element name="Etage" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="StructureVerticalePBOType" nillable="true" type="tns:StructureVerticalePBOType"/>
			<xs:complexType name="ListeFibreType">
				<xs:sequence>
					<xs:element minOccurs="0" maxOccurs="unbounded" name="Fibres" nillable="true" type="tns:FibreType"/>
				</xs:sequence>
			</xs:complexType>
			<xs:complexType name="FibreType">
				<xs:sequence>
					<xs:element name="EtatFibre" nillable="true" type="tns:EtatFibreType"/>
					<xs:element name="IdentifiantFibre" nillable="true" type="xs:string"/>
					<xs:element name="InformationFibrePBO" nillable="true" type="xs:string"/>
					<xs:element name="InformationTubePBO" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferenceCablePBO" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferencePrise" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="FibreType" nillable="true" type="tns:FibreType"/>
			<xsd:simpleType name="EtatFibreType">
				<xsd:restriction base="xsd:string">
					<xsd:enumeration value="fibre disponible"/>
					<xsd:enumeration value="fibre dédiée disponible"/>
					<xsd:enumeration value="fibre réservée"/>
					<xsd:enumeration value="fibre occupée"/>
					<xsd:enumeration value="fibre hors service"/>
				</xsd:restriction>
			</xsd:simpleType>
			
			<xs:complexType name="ReferenceAdresseDemandeType">
				<xs:sequence>
					<xs:element minOccurs="0" name="IdentifiantImmeuble" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferenceGeographique" nillable="true" type="tns:CoordonneesGeographiquesType"/>
					<xs:element minOccurs="0" name="ReferenceHexacle" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferenceHexacleVoie" nillable="true" type="tns:ReferenceHexacleVoieType"/>
					<xs:element minOccurs="0" name="ReferenceRivoli" nillable="true" type="tns:ReferenceRivoliType"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="ReferenceAdresseDemandeType" nillable="true" type="tns:ReferenceAdresseDemandeType"/>
			
			<xs:complexType name="ListeRoutesOptiques">
				<xs:sequence>
					<xs:element minOccurs="0" maxOccurs="4" name="RoutesOptiques" nillable="true" type="tns:RouteOptique"/>
				</xs:sequence>
			</xs:complexType>

			<xs:complexType name="RouteOptique">
				<xs:sequence>
					<xs:element minOccurs="0" name="ConnecteurPriseCouleur" nillable="true" type="xs:string"/>
					<xs:element name="ConnecteurPriseNumero" type="xs:int"/>
					<xs:element name="InformationFibrePBO" nillable="true" type="xs:string"/>
					<xs:element name="InformationTubePBO" nillable="true" type="xs:string"/>
					<xs:element name="OC" nillable="true" type="xs:string"/>
					<xs:element name="PositionPm" nillable="true" type="tns:PositionPmType"/>
					<xs:element minOccurs="0" name="ReferenceCablePBO" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="RouteOptique" nillable="true" type="tns:RouteOptique"/>
			<xs:complexType name="PositionPmType">
				<xs:sequence>
					<xs:element minOccurs="0" name="InfoFibreModulePM" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="InfoTubeModulePM" nillable="true" type="xs:string"/>
					<xs:element name="NomModulePM" nillable="true" type="xs:string"/>
					<xs:element name="PositionModulePM" nillable="true" type="xs:string"/>
					<xs:element minOccurs="0" name="ReferenceCableModulePM" nillable="true" type="xs:string"/>
				</xs:sequence>
			</xs:complexType>
			<xs:element name="PositionPmType" nillable="true" type="tns:PositionPmType"/>
		</xs:schema>
	</wsdl:types>
	<wsdl:message name="Emutation_RecherchePBO_InputMessage">
		<wsdl:part name="parameters" element="tns:RecherchePBO"/>
	</wsdl:message>
	<wsdl:message name="Emutation_RecherchePBO_OutputMessage">
		<wsdl:part name="parameters" element="tns:RecherchePBOResponse"/>
	</wsdl:message>
	<wsdl:message name="Emutation_ConsultationFibres_InputMessage">
		<wsdl:part name="parameters" element="tns:ConsultationFibres"/>
	</wsdl:message>
	<wsdl:message name="Emutation_ConsultationFibres_OutputMessage">
		<wsdl:part name="parameters" element="tns:ConsultationFibresResponse"/>
	</wsdl:message>
	<wsdl:message name="Emutation_MiseAJourRouteOptique_InputMessage">
		<wsdl:part name="parameters" element="tns:MiseAJourRouteOptique"/>
	</wsdl:message>
	<wsdl:message name="Emutation_MiseAJourRouteOptique_OutputMessage">
		<wsdl:part name="parameters" element="tns:MiseAJourRouteOptiqueResponse"/>
	</wsdl:message>
	<wsdl:portType name="Emutation">
		
		<wsdl:operation name="RecherchePBO">
			<wsdl:input wsaw:Action="http://interop-fibre.fr/Emutation/1.0/RecherchePBO" message="tns:Emutation_RecherchePBO_InputMessage"/>
			<wsdl:output wsaw:Action="http://interop-fibre.fr/Emutation/1.0/RecherchePBOResponse" message="tns:Emutation_RecherchePBO_OutputMessage"/>
		</wsdl:operation>
		<wsdl:operation name="ConsultationFibres">
			<wsdl:input wsaw:Action="http://interop-fibre.fr/Emutation/1.0/ConsultationFibres" message="tns:Emutation_ConsultationFibres_InputMessage"/>
			<wsdl:output wsaw:Action="http://interop-fibre.fr/Emutation/1.0/ConsultationFibresResponse" message="tns:Emutation_ConsultationFibres_OutputMessage"/>
		</wsdl:operation>
		<wsdl:operation name="MiseAJourRouteOptique">
			<wsdl:input wsaw:Action="http://interop-fibre.fr/Emutation/1.0/MiseAJourRouteOptique" message="tns:Emutation_MiseAJourRouteOptique_InputMessage"/>
			<wsdl:output wsaw:Action="http://interop-fibre.fr/Emutation/1.0/MiseAJourRouteOptiqueResponse" message="tns:Emutation_MiseAJourRouteOptique_OutputMessage"/>
		</wsdl:operation>
	</wsdl:portType>
	<wsdl:binding name="BasicHttpBinding_Emutation" type="tns:Emutation">
		<soap:binding transport="http://schemas.xmlsoap.org/soap/http"/>
		<wsdl:operation name="RecherchePBO">
			<soap:operation soapAction="http://interop-fibre.fr/Emutation/1.0/RecherchePBO" style="document"/>
			<wsdl:input>
				<soap:body use="literal"/>
			</wsdl:input>
			<wsdl:output>
				<soap:body use="literal"/>
			</wsdl:output>
		</wsdl:operation>
		<wsdl:operation name="ConsultationFibres">
			<soap:operation soapAction="http://interop-fibre.fr/Emutation/1.0/ConsultationFibres" style="document"/>
			<wsdl:input>
				<soap:body use="literal"/>
			</wsdl:input>
			<wsdl:output>
				<soap:body use="literal"/>
			</wsdl:output>
		</wsdl:operation>
		<wsdl:operation name="MiseAJourRouteOptique">
			<soap:operation soapAction="http://interop-fibre.fr/Emutation/1.0/MiseAJourRouteOptique" style="document"/>
			<wsdl:input>
				<soap:body use="literal"/>
			</wsdl:input>
			<wsdl:output>
				<soap:body use="literal"/>
			</wsdl:output>
		</wsdl:operation>
	</wsdl:binding>
	<wsdl:service name="Emutation">
		<wsdl:port name="BasicHttpBinding_Emutation" binding="tns:BasicHttpBinding_Emutation">
			<soap:address location="http://interop-fibre.fr/Emutation/1.0/Emutation"/>
		</wsdl:port>
	</wsdl:service>
</wsdl:definitions>
```

aminvielledebat replied:
Ok. Déjà, en faisant un check rapide, je vois que tu as décrit un attribut ```element``` sur tes éléments de requêtes et de réponse. Cet attribut n'est pas autorisé.
Pas dit que ça aide. Tu peux essayer sans ?

Ex : 
![image](https://user-images.githubusercontent.com/37697258/40780812-055a766a-64da-11e8-9439-9a522c15a1ec.png)


Pour ce qui est de l'appli en C#, on parle de vos devs ou de l'outil utilisé pour le requêtage et/ou la génération.
J'ai pas de stacks chez moi, mais en donnant le nom des produits, ça peut peut être aider.

dlevasseuraltitude replied:
Alors on parle de mon appli en C# qui appelle le wsdl. J'ai aussi essayé sans les attributs <code>element</code> mais ça n'a pas donné grand chose de plus

aminvielledebat replied:
Tu penses que tu pourrais nous montrer la stack / exception ?

Sinon tu penses que tu pourrais nous donner les librairies utiliser, un squelette, ou quelque chose pour que l'on intègre dans le ```.travis.yml``` ? Ainsi on serait sûr que le client généré en C# fonctionne (ou pas), et de te proposer une correction sans avoir à monter un environnement (via travis donc).

Pour la doc si tu veux : https://docs.travis-ci.com/user/languages/csharp/. Tu peux aussi monter ça sur ton fork si tu ne souhaites pas l'afficher ici.

dlevasseuraltitude replied:
Je vais expliquer un peu plus mon problème. Je n'ai pas d'exception à l'import du wsdl, Visual Studio crée bien les classes etc... sauf qu'il me manque des types, notamment les ListePBOType, ListeFibresType. De plus, les types de retours des méthode sont faux. Pour toutes les méthodes, le type de retour est CodeRetourType, alors que ce devrait être un RecherchePBOReponse par exemple. Avec l'encapsulation, je n'ai pas ce problème.

Au niveau des librairies utilisées, ce sont les librairies System de base, mais j'ai aussi des librairies que nous avons développés.

aminvielledebat replied:
Ahhh ! Alors là, comme ça, je me dis que les deux définitions de schemas, avec les tns pas forcément forcés, ça ne doit pas arranger les choses.

Est-ce que tu peux essayer avec ce wsdl stp ?

[mutation.wsdl.txt](https://github.com/before-interop/emutation/files/2058726/mutation.wsdl.txt)

dlevasseuraltitude replied:
J'ai essayé avec ton wsdl, mais j'ai le même problème. Les types de retour des méthodes sont toujours des CodeRetourType.

Avec ton wsdl: 

```
[System.Web.Services.Protocols.SoapDocumentMethodAttribute("http://interop-fibre.fr/Emutation/1.0/ConsultationFibres", RequestNamespace="http://interop-fibre.fr/Emutation/1.0/", ResponseNamespace="http://interop-fibre.fr/Emutation/1.0/", Use=System.Web.Services.Description.SoapBindingUse.Literal, ParameterStyle=System.Web.Services.Protocols.SoapParameterStyle.Wrapped)]
        [return: System.Xml.Serialization.XmlElementAttribute("CodeRetour", IsNullable=true)]
        public CodeRetourType ConsultationFibres([System.Xml.Serialization.XmlElementAttribute(IsNullable=true)] EnteteRequeteType Entete, [System.Xml.Serialization.XmlElementAttribute(IsNullable=true)] string ReferenceCommandePriseInterneOC, [System.Xml.Serialization.XmlElementAttribute(IsNullable=true)] string ReferencePBO, [System.Xml.Serialization.XmlElementAttribute(IsNullable=true)] string ReferencePM, [System.Xml.Serialization.XmlElementAttribute(IsNullable=true)] string ReferencePrestationPrise, [System.Xml.Serialization.XmlElementAttribute("Entete", IsNullable=true)] out EnteteReponseType Entete1, [System.Xml.Serialization.XmlArrayAttribute(IsNullable=true)] [System.Xml.Serialization.XmlArrayItemAttribute("Fibres")] out FibreType[] Fibre, [System.Xml.Serialization.XmlElementAttribute(IsNullable=true)] out PboType PboType) {
            object[] results = this.Invoke("ConsultationFibres", new object[] {
                        Entete,
                        ReferenceCommandePriseInterneOC,
                        ReferencePBO,
                        ReferencePM,
                        ReferencePrestationPrise});
            Entete1 = ((EnteteReponseType)(results[1]));
            Fibre = ((FibreType[])(results[2]));
            PboType = ((PboType)(results[3]));
            return ((CodeRetourType)(results[0]));
        }
```

Avec l'encapsulation:

```
        [System.Web.Services.Protocols.SoapDocumentMethodAttribute("http://interop-fibre.fr/Emutation/1.0/ConsultationFibres", RequestNamespace="http://interop-fibre.fr/Emutation/1.0/", ResponseNamespace="http://interop-fibre.fr/Emutation/1.0/", Use=System.Web.Services.Description.SoapBindingUse.Literal, ParameterStyle=System.Web.Services.Protocols.SoapParameterStyle.Wrapped)]
        public ConsultationFibresPBOReponse ConsultationFibres(ConsultationFibresDemande demande) {
            object[] results = this.Invoke("ConsultationFibres", new object[] {
                        demande});
            return ((ConsultationFibresPBOReponse)(results[0]));
        }
```

aminvielledebat replied:
Aie. Effectivement, on dirait qu'il renvoie le premier élément compris dans le type complexe et pas l'intégralité.
La lib pour ton générateur, c'est du standard ou du custom ? (histoire que je sache où regarder parce que ça serait dommage de passer à côté)

dlevasseuraltitude replied:
C'est du standard, c'est VisualStudio qui génère en auto les classes à partir du wsdl

aminvielledebat replied:
Ok. Je viens de faire le test avec wsdl.exe et svcutils.exe et effectivement, il manque bien des choses (moins avec svcutils mais quand même).

Je suis tombé sur plusieurs sujets qui remontent le problème, et apparemment, on arrive à le corriger en modifiant le wsdl. Je cherche.

aminvielledebat replied:
Ok, j'ai trouvé.... https://developers.de/blogs/andreas_erben/archive/2007/02/02/svcutil.exe_2F00_wsdl.exe_3A00_-issue-with-message-part-name-_2200_parameters_2200_.aspx

Juste le temps de te faire la modif, et de te la remettre ici.

aminvielledebat replied:
Et voilà [emutation.wsdl.txt](https://github.com/before-interop/emutation/files/2062452/emutation.wsdl.txt)

Peux-tu me dire si c'est mieux ?
Si ça passe chez toi, il faudrait rebase (ou merge, je ne vais pas imposer non plus :-)) ta branche sur le master, pour que l'on voit si la génération sur les différents langages fonctionne.

dlevasseuraltitude replied:
Alors oui c'est mieux je génère bien les types, par contre on a un écart par rapport à la doc. Les paramètres d'entrées avec cette version se nomment avec le même nom que la méthode, par exemple "RecherchePBO", alors que dans la doc le paramètre s'appelle "RechechePBODemande". De même pour la réponse, le type est "RecherchePBOResponse" alors que ce devrait être "RecherchePBOReponse". Mais je ne sais pas que si c'est possible de faire ça sans encapsulation

aminvielledebat replied:
Cool ! En voilà une bonne nouvelle :smile: 

Pour cette histoire de nommage, je sais qu'en termes de best-practices chez nous, on fait comme tu l'as décrit à savoir ```NomOperation``` = ```NomElementRequete``` et pour la réponse, idem mais avec ```Response``` à la fin. 
Personnellement, je ne ferai pas une jaunisse si ça change, mais je pars du principe que cela fait partie de l'implémentation, sauf si on vient nous dire qu'il faut absolument respecter ce qui est défini dans la spec... et dans ce cas, ton encapsulation ne respecte pas non plus le document :smiley:

dlevasseuraltitude replied:
OK Merci du coup de main :)

Par contre, pourquoi la version avec encapsulation ne respecterai pas le document? Au niveau des type etc... tout me semblait OK

aminvielledebat replied:
Regarde les signatures, c'est plus simple : 

# La version avec encapsulation :
``` php
public function RecherchePBO(RecherchePBO $parameters): RecherchePBOResponse;
```
Avec : 
```php 
class RecherchePBO
{
    /**
     * @var RecherchePBODemande $demande
     */
    protected $demande = null;
...
}

class RecherchePBODemande
{

    /**
     * @var EnteteRequeteType $Entete
     */
    protected $Entete = null;

    /**
     * @var ReferenceCommandePriseInterneOC $ReferenceCommandePriseInterneOC
     */
    protected $ReferenceCommandePriseInterneOC = null;

    /**
     * @var ReferencePrestationPrise $ReferencePrestationPrise
     */
    protected $ReferencePrestationPrise = null;

...
}
```
Le xml associé : 
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns="http://interop-fibre.fr/Emutation/1.0/">
   <soapenv:Header/>
   <soapenv:Body>
      <ns:ConsultationFibres>
         <!--Optional:-->
         <ns:demande>
            <ns:Entete>
               <ns:HorodatageRequete>?</ns:HorodatageRequete>
...
```

# La version sans encapsulation : 

``` php 
public function RecherchePBO(RecherchePBO $request): RecherchePBOResponse;
```

Avec : 
```php 
class RecherchePBO
{

    /**
     * @var EnteteRequeteType $Entete
     */
    protected $Entete = null;

    /**
     * @var ReferenceCommandePriseInterneOC $ReferenceCommandePriseInterneOC
     */
    protected $ReferenceCommandePriseInterneOC = null;

    /**
     * @var ReferencePrestationPrise $ReferencePrestationPrise
     */
    protected $ReferencePrestationPrise = null;
...
}
```

Le xml associé : 
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ns="http://interop-fibre.fr/Emutation/1.0/">
   <soapenv:Header/>
   <soapenv:Body>
      <ns:RecherchePBO>
         <ns:Entete>
            <ns:HorodatageRequete>?</ns:HorodatageRequete>
...
```

# La spec qui va bien : 

![image](https://user-images.githubusercontent.com/37697258/40845336-b2bd5764-65b6-11e8-968a-7bca9fa714fb.png)

aminvielledebat replied:
Dans le même sens si on pousse le bouchon, la réponse ne respecte pas la spec non plus : 

Le wsdl : 
```xml
            <xsd:element name="RecherchePBOResponse">
                <xsd:complexType>
                    <xsd:sequence>
                        <xsd:element name="CodeRetour" nillable="true" type="tns:CodeRetourType"/>
                        <xsd:element name="Entete" nillable="true" type="tns:EnteteReponseType"/>
                        <xsd:element name="ListePBO" nillable="true" type="tns:ListePboType"/>
                    </xsd:sequence>
                </xsd:complexType>
            </xsd:element>
```

La spec : 
![image](https://user-images.githubusercontent.com/37697258/40845487-1fb4816c-65b7-11e8-84df-dc8bf97e1486.png)

On voit bien un delta sur l'ordre des paramètres ainsi que sur la case de ListePBO... mais je pense que l'on peut dire que c'est de l'implémentation non ou on va juste qu'au bout du bout ?

aminvielledebat replied:
Géré dans la #21

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+			<xs:attribute name="Ref" type="xs:IDREF"/>
+		</xs:schema>
+		<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:ser="http://schemas.microsoft.com/2003/10/Serialization/" xmlns:tns="http://schemas.datacontract.org/2004/07/" elementFormDefault="qualified" targetNamespace="http://schemas.datacontract.org/2004/07/">
+			<xs:import namespace="http://schemas.microsoft.com/2003/10/Serialization/"/>
```
Techno microsoft ici encore. A supprimer en interop' selon moi.

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+		<wsdl:part name="parameters" element="tns:RecherchePBOResponse"/>
+	</wsdl:message>
+	<wsdl:message name="Emutation_ConsultationFibres_InputMessage">
+		<wsdl:part name="parameters" element="tns:ConsultationFibres"/>
```
Personnellement, j'ai un problème avec différents outils sur les éléments et leurs namespaces.

Ex sous SoapUI : 
![image](https://user-images.githubusercontent.com/37697258/40393178-ba5f4e36-5e1e-11e8-9973-1f1f664a5b2b.png)

Cela interdit toute génération de requêtes, ou de code en fonction des outils utilisés.
Je pense que ce n'est pas quelque chose de grave (une déclaration d'élément qu'il doit falloir faire par ci, par là).

aminvielledebat replied:
Au final, je n'ai rien dit sur ce point... En reprenant le wsdl proprement depuis ta PR, je n'ai plus de problème, va savoir pourquoi...

Ok donc pour SoapUi, et toutes mes excuses.

aminvielledebat reviewed (changes requested):
Côté Orange, après une première relecture un peu plus poussée, on constate qu'il y a encore des points bloquants : 
- Pas conforme à la spécification interop' sur le format des champs
- Trop lié à Microsoft => problème d’interopérabilité potentiel.
- Erreur un une liste d'objets qui ne décrit pas une collection
- Erreurs avec plusieurs outils de génération empêchant les autres ORTs de travailler sur un pied d'égalité.

Si besoin, n'hésitez pas à revenir vers nous.

aminvielledebat commented on emutation.wsdl:
```diff
@@ -1,401 +1,385 @@
+			<xs:import namespace="http://schemas.microsoft.com/2003/10/Serialization/"/>
+			<xs:complexType name="RecherchePBODemande">
+				<xs:sequence>
+					<xs:element name="Entete" nillable="true" type="tns:EnteteRequeteType"/>
```
L'élément ```Entete``` de devrait pas être dans le header Soap ?

Du style passer de ce format : 
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:int="http://interop-fibre.fr/" xmlns:ns="http://schemas.datacontract.org/2004/07/">
   <soapenv:Header>
      </soapenv:Header>
   <soapenv:Body>
      <int:RecherchePBO>
            <ns:Entete>
               <ns:HorodatageRequete>?</ns:HorodatageRequete>
               <ns:OperateurCommercial>
                  <!--Optional:-->
                  <ns:Identifiant>?</ns:Identifiant>
                  <ns:Nom>?</ns:Nom>
               </ns:OperateurCommercial>
               <ns:VersionWS>?</ns:VersionWS>
            </ns:Entete>
            <!--Optional:-->
            <ns:ReferenceCommandePriseInterneOC>?</ns:ReferenceCommandePriseInterneOC>
            <!--Optional:-->
            <ns:ReferencePrestationPrise>?</ns:ReferencePrestationPrise>
      </int:RecherchePBO>
   </soapenv:Body>
</soapenv:Envelope>
``` 
A ce format : 
```xml
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:int="http://interop-fibre.fr/" xmlns:ns="http://schemas.datacontract.org/2004/07/">
   <soapenv:Header>
      <int:EnteteRequeteType>
         <int:HorodatageRequete>?</int:HorodatageRequete>
         <int:OperateurCommercial>
            <!--Optional:-->
            <int:Identifiant>?</int:Identifiant>
            <int:Nom>?</int:Nom>
         </int:OperateurCommercial>
         <int:VersionWS>?</int:VersionWS>
      </int:EnteteRequeteType>
   </soapenv:Header>
   <soapenv:Body>
      <int:RecherchePBODemande>
         <!--Optional:-->
         <int:ReferenceCommandePriseInterneOC>?</int:ReferenceCommandePriseInterneOC>
         <!--Optional:-->
         <int:ReferencePrestationPrise>?</int:ReferencePrestationPrise>
      </int:RecherchePBODemande>
   </soapenv:Body>
</soapenv:Envelope>
```

dlevasseuraltitude replied:
De notre côté on a pas l'habitude de gérer ça de cette manière. Cet élément EnteteRequeteType est un paramètre de la méthode RecherchePBO. De plus, en déplaçant cet entête, je ne suis pas sûr que l'on respecte la documentation. Quel est l’inconvénient pour toi de ne pas avoir l'EnteteRequeteType  dans le header soap?

aminvielledebat replied:
Aucun inconvénient de mon côté. C'est simplement une interprétation de la spécification que j ai pu faire.
Je suis parti du principe que si l'objet s'appelait "entête", il y avait des chances que cela soit un entête :-)

Après, en y réfléchissant, je ne trouve pas absurde que ces informations là ne soient pas dans le body (on parle de paramètres techniques ou d'éléments permettant la validation de la signature du contrat, pas d'informations liées à la mutation à proprement parlé).

Pour moi, c'est le genre de sujets sur lequel les représentants techniques peuvent proposer une solution répondant au besoin de la spec. On reste conforme selon moi. Aux representants de trancher (via une autre PR)

aminvielledebat replied:
@jlsaxione un avis là dessus ? On fait une PR à côté pour sortir l'objet d'entête du body, et le mettre dans le header soap ou on reste en l'état ?

aminvielledebat replied:
tracé dans la #22

dlevasseuraltitude committed:
Update README.md

Modification du fichier README.md suite à la modification du wsdl

dlevasseuraltitude committed:
Update emutation.wsdl

Mise à jour du fichier wsdl suite aux différentes remontées

aminvielledebat commented:
@jlsaxione @dlevasseuraltitude, je propose que l'on valide les derniers échanges que l'on a pu avoir (le fichier que je t'ai envoyé, si cela te convient @dlevasseuraltitude) et que l'on fasse d'autres PRs pour la suite, sinon cela vite devenir illisible.

De mon côté, j'en prépare une autre pour la mise en conformité avec la spec (il manque encore pas mal de restrictions).

En phase avec ça ?

jlsaxione commented:
en phase avec ça même si je vous ai laissé travailler sur le sujet tous les deux.
Il faudra aussi diffuser ce wsdl, voir le tagger pour que les MOA le propage vers toutes les équipes de dev de chaque opérateur.

aminvielledebat commented:
Ping pour cette PR @dlevasseuraltitude (Ok pour Jérôme), comme ça, vous pourrez jeter un coup d'oeil à la #21 où j'ai repris les restrictions et l'ordre des champs dans les séquences, comme défini dans la spec interop'.

jlsaxione commented:
on accepte le merge ?

aminvielledebat commented:
Yes. Il faut que l'on avance, et il reste pas mal de non conformités que j'ai corrigé dans ma dernière #21.

Reste à corriger ce qui fait péter le build travis.
